### ISSUE TITLE
IPC::Open3 should suicide in the child when exec fails after fork, instead of croaking

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8471]
Migrated from [rt.perl.org#39252](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39252) (status was 'resolved')

Searchable as RT39252$

**Comment by p5pRT** [comment]
### From xdrudis@tinet.cat

#### Created by xdrudis@tinet.cat




Hello\.

I ran into surprise using IPC&#8203;::Open3 when the program to run didn't
exist\. Instead of simply failing\, the call leaves a child process
alive \(when called inside an eval\) and the rest of the program
executes twice\, once the parent and once the child\.

The child croaks the error but this is no help if called inside
an eval \(and is only useful for dying otherwise\)\. In fact the
croak in perl 5\.8\.8 has a comment suggesting it's a mistake\,
but it isn't fixed \(I hadn't noticed the comment at first\,
when I first encountered this bug on perl 5\.8\.2 on AIX\, but
I see it now on perl 5\.8\.8 on debian sid\)\.

Having consulted at a Barcelona Perl Mongers meeting\, Salvador
Fandiño \(who commented having had a similar bug in a module of his own\)
provided shorter test cases and a few solutions&#8203;:

1\- calling exit would work for the eval case\, but would still
execute the END block in the child which is unexpected\.

2\- calling POSIX&#8203;::\_exit would work if loading POSIX is not too
cumbersome

3\- the child could suicide with either exec\("/bin/false"\)

4\- or kill\(9\,$$\)\.

Here are Salvador's test cases \(I don't include mine since it's longer\)

this would be fixed \(only one pid was here message\) with either solution
\(1 to 4\) :

&nbsp;   use IPC&#8203;::Open3;
&nbsp;   eval \{ open3\("\<&STDIN"\, ">&STDOUT"\, ">&STDERR"\, "/foo/bar"\) \};
&nbsp;   print "$$ was here\\n";

this would only be fixed by solutions 2\, 3 or 4&#8203;:

&nbsp;   use IPC&#8203;::Open3;
&nbsp;   END \{ print "END&#8203;: $$ was here\\n"; \}
&nbsp;   open3\("\<&STDIN"\, ">&STDOUT"\, ">&STDERR"\, "/foo/bar"\);


I don't know which solution is more portable I'm sending a tiny patch
for 4\. Hope it helps\.

Note that without croaking the error can still be known in the father
by checking $?\, at least in my tests\. I haven't looked at whether 
the father can check $? still in open3 in order to croak in the 
father even if the error was in the child\. Might work if necessary\.
But anyway a missing exception is less disturbing than a surviving child 
you didn't fork\. 

Thanks a lot for your work making perl what it is\.

\*\*\* IPC/Open3\.pm        2006\-05\-30 19&#8203;:03&#8203;:41\.000000000 \+0200
\-\-\- fixed/IPC/Open3\.pm  2006\-05\-30 19&#8203;:12&#8203;:25\.000000000 \+0200
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 166\,172 \*\*\*\*
&nbsp;        \}
&nbsp;        local\($"\)=\(" "\);
&nbsp;        exec @&#8203;cmd \# XXX&#8203;: wrong process to croak from
\!           or croak "$Me&#8203;: exec of @&#8203;cmd failed";
&nbsp;      \} elsif \($do\_spawn\) \{
&nbsp;        \# All the bookkeeping of coincidence between handles is
&nbsp;        \# handled in spawn\_with\_handles\.
\-\-\- 166\,172 \-\-\-\-
&nbsp;        \}
&nbsp;        local\($"\)=\(" "\);
&nbsp;        exec @&#8203;cmd \# XXX&#8203;: wrong process to croak from
\!           or kill\(9\,$$\);
&nbsp;      \} elsif \($do\_spawn\) \{
&nbsp;        \# All the bookkeeping of coincidence between handles is
&nbsp;        \# handled in spawn\_with\_handles\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium

Site configuration information for perl v5.8.8:

Configured by Debian Project at Tue Apr  4 22:34:25 UTC 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.15.4, archname=i486-linux-gnu-thread-multi
    uname='linux ninsei 2.6.15.4 #1 smp preempt mon feb 20 09:48:53 pst 2006 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.8 -Dsitearch=/usr/local/lib/perl/5.8.8 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.8 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.0.3 (Debian 4.0.3-1)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.6.so, so=so, useshrplib=true, libperl=libperl.so.5.8.8
    gnulibc_version='2.3.6'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /etc/perl
    /usr/local/lib/perl/5.8.8
    /usr/local/share/perl/5.8.8
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.8:
    HOME=/home/xdrudis
    LANG=ca_ES@euro
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, May 30\, 2006 at 10&#8203;:22&#8203;:36PM \-0700\, xdrudis @&#8203; tinet\. cat wrote&#8203;:
> I ran into surprise using IPC&#8203;::Open3 when the program to run didn't
> exist\. Instead of simply failing\, the call leaves a child process
> alive \(when called inside an eval\) and the rest of the program
> executes twice\, once the parent and once the child\.

funnily enough I developed a patch for this exact problem a few days ago\.
I will apply it to the devlopment branch of perl in a day or two\.

\-\- 
"The GPL violates the U\.S\. Constitution\, together with copyright\,
antitrust and export control laws"
&nbsp;    \-\- SCO smoking crack again\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, May 31\, 2006 at 12&#8203;:33&#8203;:25PM \+0100\, Dave Mitchell wrote&#8203;:
> On Tue\, May 30\, 2006 at 10&#8203;:22&#8203;:36PM \-0700\, xdrudis @&#8203; tinet\. cat wrote&#8203;:
> > I ran into surprise using IPC&#8203;::Open3 when the program to run didn't
> > exist\. Instead of simply failing\, the call leaves a child process
> > alive \(when called inside an eval\) and the rest of the program
> > executes twice\, once the parent and once the child\.
> 
> funnily enough I developed a patch for this exact problem a few days ago\.
> I will apply it to the devlopment branch of perl in a day or two\.

Now fixed in bleedperl by change \#28347

\-\- 
Now is the discount of our winter tent
&nbsp;    \-\- sign seen outside camping shop

**Comment by p5pRT** [comment]
### From @salva

Dave Mitchell wrote&#8203;:
> On Wed\, May 31\, 2006 at 12&#8203;:33&#8203;:25PM \+0100\, Dave Mitchell wrote&#8203;:
>> On Tue\, May 30\, 2006 at 10&#8203;:22&#8203;:36PM \-0700\, xdrudis @&#8203; tinet\. cat wrote&#8203;:
>>> I ran into surprise using IPC&#8203;::Open3 when the program to run didn't
>>> exist\. Instead of simply failing\, the call leaves a child process
>>> alive \(when called inside an eval\) and the rest of the program
>>> executes twice\, once the parent and once the child\.
>> funnily enough I developed a patch for this exact problem a few days ago\.
>> I will apply it to the devlopment branch of perl in a day or two\.
> 
> Now fixed in bleedperl by change \#28347
> 

calling exit doesn't solve the second part of the bug&#8203;:

&nbsp;   use IPC&#8203;::Open3;
&nbsp;   END \{ print "END&#8203;: $$ was here\\n"; \}
&nbsp;   open3\("\<&STDIN"\, ">&STDOUT"\, ">&STDERR"\, "/foo/bar"\);

END handlers are executed from the child process when exit is called\. 
The right solution is to call POSIX&#8203;::\_exit instead\.

Cheers\,

&nbsp;   \- Salvador

**Comment by p5pRT** [comment]
### From @rgarcia

On 02/06/06\, Salvador Fandiño \<sfandino@&#8203;yahoo\.com> wrote&#8203;:
> calling exit doesn't solve the second part of the bug&#8203;:
>
>    use IPC&#8203;::Open3;
>    END \{ print "END&#8203;: $$ was here\\n"; \}
>    open3\("\<&STDIN"\, ">&STDOUT"\, ">&STDERR"\, "/foo/bar"\);
>
> END handlers are executed from the child process when exit is called\.
> The right solution is to call POSIX&#8203;::\_exit instead\.

I've modified IPC&#8203;::Open3 to call \_exit instead as change \#28350\.
